File: packages/python-llm-conversation/src/mightydatainc_llm_conversation/llm_conversation.py
import json
import re
from typing import Any, Callable

from .llm_providers import get_model_name, identify_llm_provider
from .llm_submit import llm_submit


class LLMConversation(list[dict[str, Any]]):
    """Stateful conversation wrapper around provider clients.

    `LLMConversation` extends `list` so message history can be indexed,
    iterated, and serialized directly. Helper methods cover the full lifecycle:
    appending role-specific messages, submitting to the provider, and reading
    the latest assistant reply in string or dict form.
    """

    def __init__(
        self,
        ai_client: Any | None = None,
        messages: list[dict[str, Any]] | None = None,
        model: str | None = None,
    ):
        """Create a conversation with optional initial history and defaults.

        Args:
            ai_client: Provider client object used by `llm_submit`.
            messages: Optional initial message history.
            model: Optional default model for future `submit` calls.
        """
        super().__init__(messages or [])
        self.ai_client = ai_client
        self.model = model
        self.last_reply: Any = None

    @property
    def llm_provider(self) -> str | None:
        """Return provider name inferred from `ai_client`, or None."""
        if self.ai_client is None:
            return None
        return identify_llm_provider(self.ai_client)

    def assign_messages(self, messages: list[dict[str, Any]] | None = None):
        """Replace conversation history with `messages` and return self."""
        self.clear()
        if messages:
            self.extend(messages)
        return self

    def clone(self):
        """Deep clone this conversation, preserving client/model references.

        Message history and structured `last_reply` values are deep-copied to
        avoid shared mutable references between instances.
        """
        cloned = LLMConversation(ai_client=self.ai_client, model=self.model)

        if self.last_reply is None or isinstance(self.last_reply, str):
            cloned.last_reply = self.last_reply
        else:
            cloned.last_reply = json.loads(json.dumps(self.last_reply))

        cloned.assign_messages(self.to_dict_list())
        return cloned

    def submit(
        self,
        message: str | dict[str, Any] | None = None,
        role: str | None = None,
        model: str | None = None,
        json_response: bool | dict[str, Any] | str | None = None,
        shotgun: int | None = None,
        warning_callback: Callable[[str], None] | None = None,
    ) -> str | dict[str, Any] | list[Any] | int | float | bool | None:
        """Submit this conversation and append the assistant reply.

        Optionally appends a message first, resolves an effective model from
        per-call override -> conversation default -> provider smart default,
        then delegates to `llm_submit`.

        Args:
            message: Optional pre-submit message string or role/content object.
            role: Optional role override for the appended message.
            model: Optional per-call model override.
            json_response: JSON mode toggle or structured schema config.
            shotgun: Parallel worker count when using shotgunning.
            warning_callback: Optional callback for retry/non-fatal warnings.

        Returns:
            The assistant reply as text or parsed JSON value, depending on
            `json_response` mode.

        Raises:
            ValueError: If `ai_client` is not set.
        """
        if self.ai_client is None:
            raise ValueError("AI client is not set. Please provide an AI client.")

        if message:
            message_obj: dict[str, Any] = {
                "role": "user",
                "content": "",
            }
            if isinstance(message, str):
                message_obj["content"] = message
            else:
                message_obj.update(message)

            if role is not None:
                message_obj["role"] = role

            self.add_message(message_obj["role"], message_obj["content"])

        selected_model = (
            model
            or self.model
            or get_model_name(self.llm_provider or "openai", "smart")
        )

        llm_reply = llm_submit(
            messages=self.to_dict_list(),
            ai_client=self.ai_client,
            model=selected_model,
            json_response=json_response,
            shotgun=shotgun,
            warning_callback=warning_callback,
        )

        self.add_assistant_message(llm_reply)
        return llm_reply

    def add_message(self, role: str, content: Any):
        """Append a message with role-aware content normalization.

        Strings are kept as-is, lists are preserved (multimodal payloads), dicts
        are pretty-JSON serialized, and all other values are coerced to string.
        """
        normalized_content: str | list[Any]
        if isinstance(content, str):
            normalized_content = content
        elif isinstance(content, list):
            normalized_content = content
        elif isinstance(content, dict):
            normalized_content = json.dumps(content, indent=2)
        else:
            normalized_content = f"{content}"

        self.append({"role": role, "content": normalized_content})
        return self

    def add_user_message(self, content: Any):
        """Append a `user` message and return self."""
        return self.add_message("user", content)

    def add_assistant_message(self, content: Any):
        """Append an `assistant` message, update `last_reply`, and return self."""
        self.last_reply = content
        return self.add_message("assistant", content)

    def add_system_message(self, content: Any):
        """Append a `system` message and return self."""
        return self.add_message("system", content)

    def add_developer_message(self, content: Any):
        """Append a `developer` message and return self."""
        return self.add_message("developer", content)

    def add_image(self, role: str, text: str, image_data_url: str):
        """Append a multimodal text+image message for the active provider.

        OpenAI receives `input_text` and `input_image` content blocks.
        Anthropic receives `image` + `text` blocks and accepts either an HTTP(S)
        URL or a base64 data URL for the image source.
        """
        llm_provider = self.llm_provider
        if not llm_provider:
            raise ValueError(
                "LLM provider cannot be identified. Please set an AI client with a supported provider."
            )

        if llm_provider == "openai" or llm_provider == "deepseek":
            self.append(
                {
                    "role": role,
                    "content": [
                        {
                            "type": "input_text",
                            "text": text,
                        },
                        {
                            "type": "input_image",
                            "image_url": image_data_url,
                            "detail": "high",
                        },
                    ],
                }
            )
            return self

        if llm_provider == "anthropic":
            trimmed_image_data_url = image_data_url.strip()
            image_source: dict[str, str]

            if re.match(r"^https?://", trimmed_image_data_url, re.IGNORECASE):
                image_source = {
                    "type": "url",
                    "url": trimmed_image_data_url,
                }
            else:
                data_url_match = re.match(
                    r"^data:(image\/[a-zA-Z0-9.+-]+);base64,([\s\S]+)$",
                    trimmed_image_data_url,
                    re.IGNORECASE,
                )
                if not data_url_match:
                    raise ValueError(
                        "Anthropic image input must be an HTTP(S) URL or a base64 data URL, such as data:image/png;base64,..."
                    )

                media_type = data_url_match.group(1).lower()
                if media_type == "image/jpg":
                    media_type = "image/jpeg"

                if media_type not in [
                    "image/jpeg",
                    "image/png",
                    "image/gif",
                    "image/webp",
                ]:
                    raise ValueError(
                        "Unsupported Anthropic image media type: "
                        + f"{media_type}. Supported types: image/jpeg, image/png, image/gif, image/webp."
                    )

                image_source = {
                    "type": "base64",
                    "media_type": media_type,
                    "data": re.sub(r"\s+", "", data_url_match.group(2)),
                }

            self.append(
                {
                    "role": role,
                    "content": [
                        {
                            "type": "image",
                            "source": image_source,
                        },
                        {
                            "type": "text",
                            "text": text,
                        },
                    ],
                }
            )
            return self

        raise ValueError(
            f"LLM provider {llm_provider} does not support image messages."
        )

    def submit_message(self, role: str, content: Any):
        """Append a message with `role`, then submit the conversation."""
        self.add_message(role, content)
        return self.submit()

    def submit_user_message(self, content: Any):
        """Append a `user` message, then submit the conversation."""
        self.add_user_message(content)
        return self.submit()

    def submit_assistant_message(self, content: Any):
        """Append an `assistant` message, then submit the conversation."""
        self.add_assistant_message(content)
        return self.submit()

    def submit_system_message(self, content: Any):
        """Append a `system` message, then submit the conversation."""
        self.add_system_message(content)
        return self.submit()

    def submit_developer_message(self, content: Any):
        """Append a `developer` message, then submit the conversation."""
        self.add_developer_message(content)
        return self.submit()

    def submit_image(self, role: str, text: str, image_data_url: str):
        """Append an image message, then submit the conversation."""
        self.add_image(role, text, image_data_url)
        return self.submit()

    def get_last_message(self):
        """Return the last message dict, or None when history is empty."""
        return self[-1] if self else None

    def get_messages_by_role(self, role: str):
        """Return all messages whose `role` exactly matches the input."""
        return [message for message in self if message.get("role") == role]

    def get_last_reply_str(self) -> str:
        """Return `last_reply` when it is a string; otherwise return empty string."""
        return self.last_reply if isinstance(self.last_reply, str) else ""

    def get_last_reply_dict(self) -> dict[str, Any]:
        """Return a deep-cloned dict version of `last_reply`, or `{}`.

        Non-dict values (including lists/primitives) return an empty dict.
        """
        try:
            cloned_reply = json.loads(json.dumps(self.last_reply))
            if isinstance(cloned_reply, dict):
                return cloned_reply
            return {}
        except Exception:
            return {}

    def get_last_reply_dict_field(self, field_name: str, default: Any = None):
        """Return one field from `last_reply` dict, or `default` when missing."""
        try:
            last_reply_dict = self.get_last_reply_dict()
            return last_reply_dict.get(field_name, default)
        except Exception:
            return default

    def to_dict_list(self) -> list[dict[str, Any]]:
        """Return a deep-copied plain list representation of message history."""
        return json.loads(json.dumps(list(self)))

File: packages/typescript-llm-conversation/src/llmConversation.ts
import {
  AIClientLike,
  getModelName,
  identifyLLMProvider,
  LLMProviderName,
} from './index.js';
import { llmSubmit } from './llmSubmit.js';

/**
 * A single message in a conversation, with a role (e.g. `"user"`,
 * `"assistant"`, `"system"`, `"developer"`) and a content value that is
 * either a plain string or an array of content parts (used for multimodal
 * messages such as images).
 */
export interface ConversationMessage {
  role: string;
  content: string | unknown[];
}

/**
 * Per-call options for {@link LLMConversation.submit}.
 *
 * @property model - Overrides the conversation's default model for this call
 *   only.
 * @property jsonResponse - When `true`, requests a plain JSON object response.
 *   When an object, that object is forwarded to the model as a JSON schema.
 * @property shotgun - When greater than 1, fires this many parallel requests
 *   and reconciles the results. See {@link gptSubmit} for details.
 */
export interface SubmitOptions {
  model?: string;
  jsonResponse?: boolean | Record<string, unknown>;
  shotgun?: number;
}

/**
 * A stateful conversation wrapper around the OpenAI Responses API.
 *
 * `LLMConversation` extends `Array<ConversationMessage>` so the message
 * history can be iterated, spread, and indexed directly. Helper methods cover
 * the full lifecycle: adding messages, submitting to the API, and reading back
 * the last reply in various forms.
 *
 * @example
 * ```ts
 * const convo = new LLMConversation([], { openaiClient: client });
 * await convo.submitUserMessage('Hello!');
 * console.log(convo.getLastReplyStr());
 * ```
 */
export class LLMConversation extends Array<ConversationMessage> {
  static get [Symbol.species](): ArrayConstructor {
    return Array;
  }

  #aiClient?: AIClientLike;
  #model?: string;
  #lastReply: unknown = null;

  get aiClient(): AIClientLike | undefined {
    return this.#aiClient;
  }

  set aiClient(value: AIClientLike | undefined) {
    this.#aiClient = value;
  }

  get model(): string | undefined {
    return this.#model;
  }

  set model(value: string | undefined) {
    this.#model = value;
  }

  get lastReply(): unknown {
    return this.#lastReply;
  }

  set lastReply(value: unknown) {
    this.#lastReply = value;
  }

  get LLMProvider(): LLMProviderName | undefined {
    if (!this.#aiClient) {
      return undefined;
    }
    return identifyLLMProvider(this.#aiClient);
  }

  /**
   * Creates a new `LLMConversation`, optionally pre-populated with messages
   * and configured with an OpenAI client and model.
   *
   * @param messages - Initial conversation history.
   * @param options - Optional client and model defaults.
   */
  constructor(
    aiClient?: AIClientLike,
    messages?: ConversationMessage[],
    model?: string
  ) {
    super(...(messages || []));
    this.#aiClient = aiClient;
    this.#model = model;
  }

  /**
   * Replaces the entire message history with `messages`, clearing the array
   * first. If `messages` is empty or undefined, the conversation is left
   * empty.
   *
   * @param messages - The new message history to assign.
   * @returns `this`, for chaining.
   */
  assignMessages(messages?: ConversationMessage[]): this {
    this.length = 0;
    if (messages?.length) {
      this.push(...messages);
    }
    return this;
  }

  /**
   * Returns a deep clone of this conversation, including all messages,
   * member values, and the `openaiClient` reference.
   *
   * @returns A new `LLMConversation` instance with the same history and
   *   configuration.
   */
  clone(): LLMConversation {
    const retval = new LLMConversation();
    retval.aiClient = this.aiClient;
    retval.model = this.model;

    if (
      this.lastReply === undefined ||
      this.lastReply === null ||
      typeof this.lastReply === 'string'
    ) {
      retval.lastReply = this.lastReply;
    } else {
      // For objects and arrays, perform a deep copy to avoid shared references.
      retval.lastReply = JSON.parse(JSON.stringify(this.lastReply));
    }

    retval.assignMessages(this.toDictList());
    return retval;
  }

  /**
   * Submits the conversation history to the OpenAI API. Automatically appends
   * the response to the conversation as an `"assistant"` message and updates
   * `lastReply` to the new content.
   *
   * Optionally appends a message to the conversation, then submits the full
   * history to the OpenAI API and appends the assistant's reply.
   *
   * When `message` is a `Record` with `jsonResponse` key, it is interpreted as a
   * JSON schema format descriptor and used as `jsonResponse`. When it has a
   * `role` key, that role is used for the message. When it has a `content` key,
   * only that content is added to the history.
   *
   * @param message - An optional message to append before submitting. May be
   *   a plain string, a content record, or omitted to submit the existing
   *   history as-is.
   * @param role - The role to use when appending `message`. Defaults to
   *   `"user"`, or the role specified in the message record if available.
   * @param options - Per-call overrides for model, JSON response format, and
   *   shotgun parallelism.
   * @returns The assistant's reply — a `string` for plain-text responses or a
   *   parsed JSON value when `jsonResponse` is set.
   * @throws {Error} If `openaiClient` is not set.
   */
  async submit(
    message?: string | Record<string, unknown>,
    role?: string,
    options: SubmitOptions = {}
  ): Promise<unknown> {
    if (!this.aiClient) {
      throw new Error('AI client is not set. Please provide an AI client.');
    }

    if (message) {
      let messageObj = {
        role: 'user',
        content: '',
      };
      if (typeof message === 'string') {
        messageObj.content = message;
      } else {
        messageObj = { ...messageObj, ...message };
      }

      if (role) {
        messageObj.role = role;
      }
      this.addMessage(messageObj.role, messageObj.content);
    }

    const model =
      options.model ||
      this.model ||
      getModelName(this.LLMProvider || 'openai', 'smart');
    let jsonResponse = options.jsonResponse;

    const llmReply = await llmSubmit(this.toDictList(), this.aiClient, {
      jsonResponse,
      model,
      shotgun: options.shotgun,
    });

    this.addAssistantMessage(llmReply);
    return llmReply;
  }

  /**
   * Appends a message with the given role to the conversation history.
   *
   * Non-string content is normalised: arrays are stored as-is, plain objects
   * are JSON-stringified, and all other values are coerced via `String()`.
   *
   * @param role - The role of the message (e.g. `"user"`, `"assistant"`).
   * @param content - The message content.
   * @returns `this`, for chaining.
   */
  addMessage(role: string, content: unknown): this {
    let normalizedContent: string | unknown[];
    if (typeof content === 'string') {
      normalizedContent = content;
    } else if (Array.isArray(content)) {
      normalizedContent = content;
    } else {
      normalizedContent = JSON.stringify(content, null, 2);
    }

    this.push({ role, content: normalizedContent });
    return this;
  }

  /**
   * Appends a `"user"` message to the conversation history.
   *
   * @param content - The message content.
   * @returns `this`, for chaining.
   */
  addUserMessage(content: unknown): this {
    return this.addMessage('user', content);
  }

  /**
   * Appends an `"assistant"` message to the conversation history.
   * Updates `lastReply` to the new content.
   *
   * @param content - The message content.
   * @returns `this`, for chaining.
   */
  addAssistantMessage(content: unknown): this {
    this.lastReply = content;
    return this.addMessage('assistant', content);
  }

  /**
   * Appends a `"system"` message to the conversation history.
   *
   * @param content - The message content.
   * @returns `this`, for chaining.
   */
  addSystemMessage(content: unknown): this {
    return this.addMessage('system', content);
  }

  /**
   * Appends a `"developer"` message to the conversation history.
   *
   * @param content - The message content.
   * @returns `this`, for chaining.
   */
  addDeveloperMessage(content: unknown): this {
    return this.addMessage('developer', content);
  }

  /**
   * Appends a multimodal message containing both a text prompt and an image
   * to the conversation history.
   *
   * @param role - The role of the message (e.g. `"user"`).
   * @param text - The text prompt accompanying the image.
   * @param imageDataUrl - A data URL (e.g. `data:image/png;base64,...`) or
   *   HTTP URL of the image to include.
   * @returns `this`, for chaining.
   */
  addImage(role: string, text: string, imageDataUrl: string): this {
    const llmProvider = this.LLMProvider;
    if (!llmProvider) {
      throw new Error(
        'LLM provider cannot be identified. Please set an AI client with a supported provider.'
      );
    }
    if (llmProvider === 'openai' || llmProvider === 'deepseek') {
      const gptMsgWithImage = {
        role: role,
        content: [
          {
            type: 'input_text',
            text: text,
          },
          {
            type: 'input_image',
            image_url: imageDataUrl,
            detail: 'high',
          },
        ],
      };
      this.push(gptMsgWithImage);
      return this;
    }

    if (llmProvider === 'anthropic') {
      const trimmedImageDataUrl = imageDataUrl.trim();

      let imageSource:
        | { type: 'url'; url: string }
        | { type: 'base64'; media_type: string; data: string };

      if (/^https?:\/\//i.test(trimmedImageDataUrl)) {
        imageSource = {
          type: 'url',
          url: trimmedImageDataUrl,
        };
      } else {
        const dataUrlMatch = trimmedImageDataUrl.match(
          /^data:(image\/[a-zA-Z0-9.+-]+);base64,([\s\S]+)$/i
        );

        if (!dataUrlMatch) {
          throw new Error(
            'Anthropic image input must be an HTTP(S) URL or a base64 data URL, such as data:image/png;base64,...'
          );
        }

        let mediaType = dataUrlMatch[1].toLowerCase();
        if (mediaType === 'image/jpg') {
          mediaType = 'image/jpeg';
        }

        if (
          !['image/jpeg', 'image/png', 'image/gif', 'image/webp'].includes(
            mediaType
          )
        ) {
          throw new Error(
            `Unsupported Anthropic image media type: ${mediaType}. Supported types: image/jpeg, image/png, image/gif, image/webp.`
          );
        }

        imageSource = {
          type: 'base64',
          media_type: mediaType,
          data: dataUrlMatch[2].replace(/\s+/g, ''),
        };
      }

      const anthropicMsgWithImage = {
        role: role,
        content: [
          {
            type: 'image',
            source: imageSource,
          },
          {
            type: 'text',
            text: text,
          },
        ],
      };
      this.push(anthropicMsgWithImage);
      return this;
    }

    throw new Error(
      `LLM provider ${llmProvider} does not support image messages.`
    );
  }

  /**
   * Appends a message with the given role, then submits the conversation.
   *
   * @param role - The role of the message to add.
   * @param content - The message content.
   * @returns The assistant's reply.
   */
  async submitMessage(role: string, content: unknown): Promise<unknown> {
    this.addMessage(role, content);
    return this.submit();
  }

  /**
   * Appends a `"user"` message, then submits the conversation.
   *
   * @param content - The user message content.
   * @returns The assistant's reply.
   */
  async submitUserMessage(content: unknown): Promise<unknown> {
    this.addUserMessage(content);
    return this.submit();
  }

  /**
   * Appends an `"assistant"` message, then submits the conversation.
   *
   * @param content - The assistant message content.
   * @returns The assistant's reply.
   */
  async submitAssistantMessage(content: unknown): Promise<unknown> {
    this.addAssistantMessage(content);
    return this.submit();
  }

  /**
   * Appends a `"system"` message, then submits the conversation.
   *
   * @param content - The system message content.
   * @returns The assistant's reply.
   */
  async submitSystemMessage(content: unknown): Promise<unknown> {
    this.addSystemMessage(content);
    return this.submit();
  }

  /**
   * Appends a `"developer"` message, then submits the conversation.
   *
   * @param content - The developer message content.
   * @returns The assistant's reply.
   */
  async submitDeveloperMessage(content: unknown): Promise<unknown> {
    this.addDeveloperMessage(content);
    return this.submit();
  }

  /**
   * Appends an image message with the given role, then submits the
   * conversation.
   *
   * @param role - The role of the message (e.g. `"user"`).
   * @param text - The text prompt accompanying the image.
   * @param imageDataUrl - A data URL or HTTP URL of the image.
   * @returns The assistant's reply.
   */
  async submitImage(
    role: string,
    text: string,
    imageDataUrl: string
  ): Promise<unknown> {
    this.addImage(role, text, imageDataUrl);
    return this.submit();
  }

  /**
   * Returns the last message in the conversation history, or `null` if the
   * conversation is empty.
   */
  getLastMessage(): ConversationMessage | null {
    return this.length ? this[this.length - 1] : null;
  }

  /**
   * Returns all messages whose `role` matches the given value.
   *
   * @param role - The role to filter by (e.g. `"user"`, `"assistant"`).
   * @returns An array of matching messages.
   */
  getMessagesByRole(role: string): ConversationMessage[] {
    return this.filter((message) => message.role === role);
  }

  /**
   * Returns the last reply as a string. Returns an empty string if the last
   * reply is not a string (e.g. a parsed JSON value).
   */
  getLastReplyStr(): string {
    return typeof this.lastReply === 'string' ? this.lastReply : '';
  }

  /**
   * Returns the last reply as a deep-cloned plain object. Returns an empty
   * object `{}` if the last reply is not a record.
   */
  getLastReplyDict(): Record<string, unknown> {
    try {
      const retval: any = JSON.parse(JSON.stringify(this.lastReply));

      // Test to make sure that the result is actually a record and not something else
      // like an array or a primitive value.
      if (
        typeof retval === 'object' &&
        retval !== null &&
        !Array.isArray(retval)
      ) {
        return retval;
      } else {
        return {};
      }
    } catch (JSONError) {
      return {};
    }
  }

  /**
   * Returns a single field from the last reply object by name.
   *
   * @param fieldName - The key to look up in the last reply object.
   * @param defaultValue - Value to return when the field is absent or the
   *   last reply is not a record. Defaults to `undefined`.
   * @returns The field value, or `defaultValue` if not found.
   */
  getLastReplyDictField(
    fieldName: string,
    defaultValue: unknown = undefined
  ): unknown {
    try {
      const lastReplyDict = this.getLastReplyDict();
      return lastReplyDict[fieldName] ?? defaultValue;
    } catch {
      return defaultValue;
    }
  }

  /**
   * Returns a deep copy of the conversation history as a plain array of
   * {@link ConversationMessage} objects, suitable for serialisation or passing
   * to {@link gptSubmit} directly.
   */
  toDictList(): ConversationMessage[] {
    return JSON.parse(JSON.stringify([...this]));
  }
}
